docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,182 @@
+
+## New APIs
+
+We add the new APIs for retreiving and interacting with nullability:
+
+```C#
+public enum Nullability
+{
+    NotApplicable,
+    NotComputed,
+    NotNull,
+    MaybeNull
+}
+
+#region Declared Nullability
+
+// These can potentially be influenced by calculated flow state, but do not change after initial declaration
+
+public interface IDiscardSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IEventSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IFieldSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface ILocalSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IMethodSymbol : ISymbol
+{
+    Nullability ReturnNullability { get; }
+    Nullability TypeArgumentsNullabilities { get; }
+    Nullability ReceiverNullability { get; }
+}
+
+public interface IParameterSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+#endregion
+
+#region Flow State Nullability
+
+public interface IArrayTypeSymbol : ITypeSymbol
+{
+    Nullability ElementNullability { get; }
+}
+
+public interface INamedTypeSymbol : ITypeSymbol
+{
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+}
+
+public interface ITypeParameterSymbol : ITypeSymbol
+{
+    Nullability ReferenceTypeConstraintNullability { get; }
+    Nullability ConstraintsNullabilities { get; }
+}
+
+public struct TypeInfo
+{
+    public Nullability Nullability { get; }
+    public Nullability ConvertedNullability { get; }
+}
+
+#endregion
+
+#region SemanticModel
+
+public abstract class SemanticModel
+{
+    public SemanticModel GetSingleStatementSemanticModel();
+}
+
+#endregion
+
+#region Printing
+
+public interface ITypeSymbol
+{
+    string ToDisplayString(Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
+    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+}
+
+/*
+ * There will be more API changes, such as on SyntaxFactory and SyntaxGenerator. This section will be updated later when we have designs for this ready.
+ */
+
+#endregion
+```
+
+## Alternatives
```

Comment:
```suggestion
## Considered Alternatives
```

(just to clarify yes, we did consider them!) But thanks for including this section, this illustrates we did go through this design.